### PR TITLE
Time additions to preprocessor

### PR DESCRIPTION
**Pull request by amoodie**
Adding "time" options to model preprocessor (i.e., rather than `--timesteps`, running `--time`, etc).

Add a few helper functions to convert between mode ltime and real-world time.

## todo

- [x] detailed documentation page (and link to it throughout) describing the intermittency factor in detail, and how the model uses this concept to scale time.
- [ ] ~create the `DeltaModel` api for timestepping for durations, rather than just number of timesteps~
- [x] create preprocessor API for working with time and timesteps and intermittency

### CODE REVIEW

**Comment by elbeejay** [comment]
Checked the time scaling math and it all works out for me :+1:

**Comment by amoodie** [comment]
just kidding -- not ready, yet. Need to complete some documentation still.

File: docs/source/guides/userguide.rst [hunk 1]
```diff
@@ -87,11 +87,45 @@ Jobs are then run with:
...
+Model simulation duration in the high-level API
+-----------------------------------------------
+
+The duration of a model run configured with the high-level API can be set up with a number of different configuration parameters.
+For more information on "time" in the model, see the complete description at :doc:`../info/modeltime`.
+
+In the high-level API, you can specify the duration to run the model by two mechanisms: 1) the number of timesteps to run the model, or 2) the duration of time to run the model.
+
+The former case is straightforward, insofar that the model determines the timestep duration and the high-level API simply iterates the model for the specified number of timestep iterations.
+To specify the number of timesteps to run the model, use the argument `--timesteps` at the command line (or `timesteps:` in the configuration YAML file).
```
Comment (elbeejay) [inline docs/source/guides/userguide.rst]: All the double-dashes (--) within code-type font render as a single long dash for me. Not sure if you think this is a problem, but just wanted to point it out

File: docs/source/guides/userguide.rst [hunk 2]
```diff
@@ -87,11 +87,45 @@ Jobs are then run with:
...
+It is also possible to specify the input run duration in units of years with the similarly named argument `--time_years` (`time_years:`).
+
+.. code:: bash
+    
+    pyDeltaRCM --config model_configuration.yml --time 31557600
+    pyDeltaRCM --config model_configuration.yml --time_years 1
+
+would each run a simulation for :math:`(86400 * 365.25)` seconds, or 1 year.
+Do not specify both time arguments, or specify time arguments with the timesteps argument.
+In the case of multiple argument specification, precedence is given in the order `timesteps` > `time` > `time_years`.
```
Comment (elbeejay) [inline docs/source/guides/userguide.rst]: Might want to considering making this a "note" or an "important" snippet - I dunno that's just my opinion but could be good to make sure people are aware of this ordering of preference.

File: docs/source/info/modeltime.rst [hunk 3]
```diff
@@ -2,8 +2,57 @@
...
+The duration of time represented by the single-value discharge assumption is also  determined by the hydrograph.
+We arrive at the so-called *intermittency factor* (:math:`I_f`) by finding the fraction of unit-time per unit-time when the river is in flood.
+For this example, the intermittency factor scales between days and years as 
+
+.. math::
+
+    \frac{25~\textrm{days}}{365.25~\textrm{days}} \approx 0.07 \equiv I_f
+
+The intermittency factor thus gives a relationship to scale pyDeltaRCM model time (which is always computed in units of seconds elapsed) from seconds to years.
+For example, for a model run that has elapsed :math:`4.32 \times 10^8` seconds and an assumed intermittency factor of 0.07:
```
Comment (elbeejay) [inline docs/source/info/modeltime.rst:38]: For whatever it is worth, this math renders kinda out of place and lower than the neighboring text (maybe due to the exponent) for me. Not sure there is any solution but just wanted to point it out.

image

File: pyDeltaRCM/model.py [hunk 4]
```diff
@@ -399,6 +399,24 @@ def SLR(self):
...
+
+            This parameter has no effect on model operations. It is used in
+            determining model run duration.
+        """
+        return self._If
+
+    @If.setter
+    def If(self, If):
+        if (If <= 0) or (If > 1):
+            raise ValueError('If must in interval (0, 1].')
```
Comment (elbeejay) [inline pyDeltaRCM/model.py]: Missing "be"?

"If must *be* in interval (0, 1]."

**Review by elbeejay — APPROVED** [review]
Looks good to me. 2 minor suggestions.

1. Add a test or two where the yaml has multiple "time" parameters to just double-check that the order of priority for the different time settings is honored

2. Either as part of the documentation, or as a warning / message in the log file it would be good to make clear that when `dt` is not a factor of `_job_end_time`, the duration of the model run will be less than the specified value in the configuration file (or command line input). As part of the documentation this could be mentioned, or if made into a warning or log message, the reduction in time could be made known to the user upon model run initiation (if verbosity is enabled). 

Other comments are more observations than actionable stuff. Documentation for the time in DeltaRCM looks really nice, example hydrograph stuff works really well I think.

File: docs/source/guides/userguide.rst (hunk 1 above)
Comment (amoodie) [reply docs/source/guides/userguide.rst]: good catch, I'll fix this.

File: docs/source/guides/userguide.rst (hunk 2 above)
Comment (amoodie) [reply docs/source/guides/userguide.rst]: good idea.

File: docs/source/info/modeltime.rst (hunk 3 above)
Comment (amoodie) [reply docs/source/info/modeltime.rst:38]: yeah, I noticed that too. I decided to just roll with it though. I guess I could switch it to just inline (e.g., 4.32e8 seconds)?

File: pyDeltaRCM/model.py (hunk 4 above)
Comment (amoodie) [reply pyDeltaRCM/model.py]: :+1:

File: docs/source/info/modeltime.rst (hunk 3 above)
Comment (elbeejay) [reply docs/source/info/modeltime.rst:38]: Sure that works, or it can stay as is, I don't really mind either way

**Comment by amoodie** [comment]
addressed comments from review, passing tests. Merge time preprocessor feature branch.

**Pull request merged**